File: modules/agent/tool_defs.go
package agent

import (
	"encoding/json"

	"github.com/woyin/OrangeCoding/modules/ai"
	"github.com/woyin/OrangeCoding/modules/tools"
)

// BuildToolDefinitions converts registered tools into provider-facing schemas.
func BuildToolDefinitions(registry *tools.ToolRegistry) []ai.ToolDefinition {
	var defs []ai.ToolDefinition
	for _, t := range registry.List() {
		params := toolParameters(t.Parameters())
		defs = append(defs, ai.ToolDefinition{
			Type: "function",
			Function: ai.FunctionDefinition{
				Name:        t.Name(),
				Description: t.Description(),
				Parameters:  params,
			},
		})
	}
	return defs
}

func toolParameters(raw json.RawMessage) ai.ToolParameter {
	var params ai.ToolParameter
	if err := json.Unmarshal(raw, &params); err != nil || params.Type == "" {
		return ai.ToolParameter{Type: "object", Properties: make(map[string]interface{})}
	}
	if params.Properties == nil {
		params.Properties = make(map[string]interface{})
	}
	return params
}

File: packages/agent/src/tool-defs.ts
/**
 * BuildToolDefinitions converts registered tools into provider-facing schemas.
 * Ported from modules/agent/tool_defs.go.
 */

import type { ToolDefinition, ToolParameter } from "@orangecoding/ai";
import { ToolRegistry } from "@orangecoding/tools";

/** BuildToolDefinitions converts registered tools into provider-facing schemas. */
export function buildToolDefinitions(registry: ToolRegistry): ToolDefinition[] {
  const defs: ToolDefinition[] = [];
  for (const t of registry.list()) {
    const params = toolParameters(t.parameters());
    defs.push({
      type: "function",
      function: {
        name: t.name(),
        description: t.description(),
        parameters: params,
      },
    });
  }
  return defs;
}

function toolParameters(raw: Record<string, unknown>): ToolParameter {
  if (!raw || !raw.type || typeof raw.type !== "string") {
    return { type: "object", properties: {} };
  }
  const params = raw as unknown as ToolParameter;
  if (!params.properties) {
    params.properties = {};
  }
  return params;
}
